File: JavaScript/test/HandGroupingTests.mjs
import { strict as assert } from 'assert';
import { Tests } from './TestFixture.mjs';
import { Enums } from '../src/Enums.mjs';
import { Card } from '../src/Card.mjs';
import { HandGrouping } from '../src/HandGrouping.mjs';

let simpleCards = [
	Card.parse('Ah'),
	Card.parse('Kd'),
	Card.parse('Qc'),
	Card.parse('Js'),
	Card.parse('Th'),
	Card.parse('9d'),
	Card.parse('8c'),
	Card.parse('7s'),
];

Tests.add('HandGrouping.constructor() null should throw', () => {
	assert.throws(() => new HandGrouping(null), /^Cards is null$/);
});

Tests.add('HandGrouping.constructor() empty should throw', () => {
	assert.throws(() => new HandGrouping([]), /^Only 5-7 cards can be grouped$/);
});

Tests.add('HandGrouping.constructor() too few cards should throw', () => {
	assert.throws(() => new HandGrouping(simpleCards.slice(0, 4)), /^Only 5-7 cards can be grouped$/);
});

Tests.add('HandGrouping.constructor() too many cards should throw', () => {
	assert.throws(() => new HandGrouping(simpleCards), /^Only 5-7 cards can be grouped$/);
});

Tests.add('HandGrouping.constructor() duplicate card should throw', () => {
	let arrange = simpleCards.slice(0, 5);
	arrange[1] = arrange[0];
	assert.throws(() => new HandGrouping(arrange), /^Duplicate cards cannot be grouped$/);
});

Tests.add('HandGrouping.constructor() invalid card should throw', () => {
	let arrange = simpleCards.slice(0, 5);
	arrange[1] = new Card();
	assert.throws(() => new HandGrouping(arrange), /^Invalid cards cannot be grouped$/);
});

Tests.add('HandGrouping.constructor() four of a kind should return it', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Ac'),
		Card.parse('As'),
		Card.parse('Kh'),
		Card.parse('Qh'),
		Card.parse('Jh'),
	]);
	assert.equal(3, actual.single.length);
	assert.equal(0, actual.pair.length);
	assert.equal(0, actual.threeOfAKind.length);
	assert.equal(1, actual.fourOfAKind.length);
	assert.equal(4, actual.fourOfAKind[0].length);
	assert.equal("Ah", actual.fourOfAKind[0][0].toString());
	assert.equal("Ad", actual.fourOfAKind[0][1].toString());
	assert.equal("Ac", actual.fourOfAKind[0][2].toString());
	assert.equal("As", actual.fourOfAKind[0][3].toString());
});

Tests.add('HandGrouping.constructor() three of a kind should return it', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Ac'),
		Card.parse('Kh'),
		Card.parse('Qh'),
		Card.parse('Jh'),
		Card.parse('Th'),
	]);
	assert.equal(4, actual.single.length);
	assert.equal(0, actual.pair.length);
	assert.equal(1, actual.threeOfAKind.length);
	assert.equal("Ah", actual.threeOfAKind[0][0].toString());
	assert.equal("Ad", actual.threeOfAKind[0][1].toString());
	assert.equal("Ac", actual.threeOfAKind[0][2].toString());
	assert.equal(0, actual.fourOfAKind.length);
});

Tests.add('HandGrouping.constructor() multiple three of a kind should return in correct order', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Ac'),
		Card.parse('Kh'),
		Card.parse('Kd'),
		Card.parse('Kc'),
		Card.parse('Th'),
	]);
	assert.equal(1, actual.single.length);
	assert.equal(0, actual.pair.length);
	assert.equal(2, actual.threeOfAKind.length);
	assert.equal(3, actual.threeOfAKind[0].length);
	assert.equal("Ah", actual.threeOfAKind[0][0].toString());
	assert.equal("Ad", actual.threeOfAKind[0][1].toString());
	assert.equal("Ac", actual.threeOfAKind[0][2].toString());
	assert.equal(3, actual.threeOfAKind[1].length);
	assert.equal("Kh", actual.threeOfAKind[1][0].toString());
	assert.equal("Kd", actual.threeOfAKind[1][1].toString());
	assert.equal("Kc", actual.threeOfAKind[1][2].toString());
	assert.equal(0, actual.fourOfAKind.length);
});

Tests.add('HandGrouping.constructor() pair should return it', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Qc'),
		Card.parse('Js'),
		Card.parse('Th'),
		Card.parse('9d'),
		Card.parse('8c'),
	]);
	assert.equal(5, actual.single.length);
	assert.equal(1, actual.single[0].length);
	assert.equal(1, actual.single[1].length);
	assert.equal(1, actual.single[2].length);
	assert.equal(1, actual.single[3].length);
	assert.equal(1, actual.single[4].length);
	assert.equal("Qc", actual.single[0][0].toString());
	assert.equal("Js", actual.single[1][0].toString());
	assert.equal(1, actual.pair.length);
	assert.equal(2, actual.pair[0].length);
	assert.equal("Ah", actual.pair[0][0].toString());
	assert.equal("Ad", actual.pair[0][1].toString());
	assert.equal(0, actual.threeOfAKind.length);
	assert.equal(0, actual.fourOfAKind.length);
});

Tests.add('HandGrouping.constructor() multiple pair should return in correct order', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Kh'),
		Card.parse('Kd'),
		Card.parse('Qh'),
		Card.parse('Qd'),
		Card.parse('Jh'),
	]);
	assert.equal(1, actual.single.length);
	assert.equal(1, actual.single[0].length);
	assert.equal("Jh", actual.single[0][0].toString());
	assert.equal(3, actual.pair.length);
	assert.equal(2, actual.pair[0].length);
	assert.equal("Ah", actual.pair[0][0].toString());
	assert.equal("Ad", actual.pair[0][1].toString());
	assert.equal(2, actual.pair[1].length);
	assert.equal("Kh", actual.pair[1][0].toString());
	assert.equal("Kd", actual.pair[1][1].toString());
	assert.equal(2, actual.pair[2].length);
	assert.equal("Qh", actual.pair[2][0].toString());
	assert.equal("Qd", actual.pair[2][1].toString());
	assert.equal(0, actual.threeOfAKind.length);
	assert.equal(0, actual.fourOfAKind.length);
});

Tests.add('HandGrouping.constructor() singles should return in correct order', () => {
	let actual = new HandGrouping(simpleCards.slice(0, 7));
	assert.equal(7, actual.single.length);
	assert.equal(1, actual.single[0].length);
	assert.equal(1, actual.single[1].length);
	assert.equal(1, actual.single[2].length);
	assert.equal(1, actual.single[3].length);
	assert.equal(1, actual.single[4].length);
	assert.equal(1, actual.single[5].length);
	assert.equal(1, actual.single[6].length);
	assert.equal("Ah", actual.single[0][0].toString());
	assert.equal("Kd", actual.single[1][0].toString());
	assert.equal("Qc", actual.single[2][0].toString());
	assert.equal("Js", actual.single[3][0].toString());
	assert.equal("Th", actual.single[4][0].toString());
	assert.equal("9d", actual.single[5][0].toString());
	assert.equal("8c", actual.single[6][0].toString());
	assert.equal(0, actual.pair.length);
	assert.equal(0, actual.threeOfAKind.length);
	assert.equal(0, actual.fourOfAKind.length);
});

Tests.add('HandGrouping.constructor() four of a kind with high cards should return correctly', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Ac'),
		Card.parse('As'),
		Card.parse('Kh'),
		Card.parse('Qh'),
		Card.parse('Jh'),
	]);
	assert.equal(3, actual.single.length);
	assert.equal(1, actual.single[0].length);
	assert.equal(1, actual.single[1].length);
	assert.equal(1, actual.single[2].length);
	assert.equal("Kh", actual.single[0][0].toString());
	assert.equal("Qh", actual.single[1][0].toString());
	assert.equal("Jh", actual.single[2][0].toString());
	assert.equal(0, actual.pair.length);
	assert.equal(0, actual.threeOfAKind.length);
	assert.equal(1, actual.fourOfAKind.length);
	assert.equal(4, actual.fourOfAKind[0].length);
	assert.equal("Ah", actual.fourOfAKind[0][0].toString());
	assert.equal("Ad", actual.fourOfAKind[0][1].toString());
	assert.equal("Ac", actual.fourOfAKind[0][2].toString());
	assert.equal("As", actual.fourOfAKind[0][3].toString());
});

Tests.add('HandGrouping.constructor() full house with high cards should return correctly', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Ac'),
		Card.parse('Kh'),
		Card.parse('Kd'),
		Card.parse('Qh'),
		Card.parse('Jh'),
	]);
	assert.equal(2, actual.single.length);
	assert.equal(1, actual.single[0].length);
	assert.equal(1, actual.single[1].length);
	assert.equal("Qh", actual.single[0][0].toString());
	assert.equal("Jh", actual.single[1][0].toString());
	assert.equal(1, actual.pair.length);
	assert.equal(2, actual.pair[0].length);
	assert.equal("Kh", actual.pair[0][0].toString());
	assert.equal("Kd", actual.pair[0][1].toString());
	assert.equal(1, actual.threeOfAKind.length);
	assert.equal(3, actual.threeOfAKind[0].length);
	assert.equal("Ah", actual.threeOfAKind[0][0].toString());
	assert.equal("Ad", actual.threeOfAKind[0][1].toString());
	assert.equal("Ac", actual.threeOfAKind[0][2].toString());
	assert.equal(0, actual.fourOfAKind.length);
});

Tests.add('HandGrouping.constructor() three of a kind with high cards should return correctly', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Ac'),
		Card.parse('Kh'),
		Card.parse('Qh'),
		Card.parse('Jh'),
		Card.parse('Th'),
	]);
	assert.equal(4, actual.single.length);
	assert.equal(1, actual.single[0].length);
	assert.equal(1, actual.single[1].length);
	assert.equal(1, actual.single[2].length);
	assert.equal(1, actual.single[3].length);
	assert.equal("Kh", actual.single[0][0].toString());
	assert.equal("Qh", actual.single[1][0].toString());
	assert.equal("Jh", actual.single[2][0].toString());
	assert.equal("Th", actual.single[3][0].toString());
	assert.equal(0, actual.pair.length);
	assert.equal(1, actual.threeOfAKind.length);
	assert.equal(3, actual.threeOfAKind[0].length);
	assert.equal("Ah", actual.threeOfAKind[0][0].toString());
	assert.equal("Ad", actual.threeOfAKind[0][1].toString());
	assert.equal("Ac", actual.threeOfAKind[0][2].toString());
	assert.equal(0, actual.fourOfAKind.length);
});

Tests.add('HandGrouping.constructor() two pair with high cards should return correctly', () => {
	let actual = new HandGrouping([
		Card.parse('Ah'),
		Card.parse('Ad'),
		Card.parse('Kh'),
		Card.parse('Kd'),
		Card.parse('Qh'),
		Card.parse('Jh'),
		Card.parse('Th'),
	]);
	assert.equal(3, actual.single.length);
	assert.equal(1, actual.single[0].length);
	assert.equal(1, actual.single[1].length);
	assert.equal(1, actual.single[2].length);
	assert.equal("Qh", actual.single[0][0].toString());
	assert.equal("Jh", actual.single[1][0].toString());
	assert.equal("Th", actual.single[2][0].toString());
	assert.equal(2, actual.pair.length);
	assert.equal(2, actual.pair[0].length);
	assert.equal("Ah", actual.pair[0][0].toString());
	assert.equal("Ad", actual.pair[0][1].toString());
	assert.equal(2, actual.pair[1].length);
	assert.equal("Kh", actual.pair[1][0].toString());
	assert.equal("Kd", actual.pair[1][1].toString());
	assert.equal(0, actual.threeOfAKind.length);
	assert.equal(0, actual.fourOfAKind.length);
});

File: CSharp/Gravical.Poker.Core.Tests/HandGroupingTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Diagnostics.CodeAnalysis;
using System.Linq;


namespace Gravical.Poker.Core.Tests
{
    [TestClass]
    [ExcludeFromCodeCoverage]
    public class HandGroupingTests : TestBase
    {
        private Card[] SimpleCards = new Card[]
        {
            new Card(Face.Ace, Suit.Heart),
            new Card(Face.King, Suit.Diamond),
            new Card(Face.Queen, Suit.Club),
            new Card(Face.Jack, Suit.Spade),
            new Card(Face.Ten, Suit.Heart),
            new Card(Face.Nine, Suit.Diamond),
            new Card(Face.Eight, Suit.Club),
            new Card(Face.Seven, Suit.Spade),
        };

        [TestMethod]
        public void Ctor_Null_ShouldThrowArgumentNullException()
        {
            ActExpectingArgumentNullException("cards", () => new HandGrouping(null));
        }

        [TestMethod]
        public void Ctor_Empty_ShouldThrowArgumentException()
        {
            ActExpectingArgumentException("cards", "Only 5-7 cards can be grouped", () => new HandGrouping(new Card[0]));
        }

        [TestMethod]
        public void Ctor_TooFewCards_ShouldThrowArgumentException()
        {
            ActExpectingArgumentException("cards", "Only 5-7 cards can be grouped", () => new HandGrouping(SimpleCards.Take(4).ToArray()));
        }

        [TestMethod]
        public void Ctor_TooManyCards_ShouldThrowArgumentException()
        {
            ActExpectingArgumentException("cards", "Only 5-7 cards can be grouped", () => new HandGrouping(SimpleCards));
        }

        [TestMethod]
        public void Ctor_DuplicateCard_ShouldThrowArgumentException()
        {
            var arrange = SimpleCards.Take(5).ToArray();
            arrange[1] = arrange[0];
            ActExpectingArgumentException("cards", "Duplicate cards cannot be grouped", () => new HandGrouping(arrange));
        }

        [TestMethod]
        public void Ctor_InvalidCard_ShouldThrowArgumentException()
        {
            var arrange = SimpleCards.Take(5).ToArray();
            arrange[1] = new Card();
            ActExpectingArgumentException("cards", "Invalid cards cannot be grouped", () => new HandGrouping(arrange));
        }

        [TestMethod]
        public void Ctor_FourOfAKind_ShouldReturnIt()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.Ace, Suit.Club),
                new Card(Face.Ace, Suit.Spade),
                new Card(Face.King, Suit.Heart),
                new Card(Face.Queen, Suit.Heart),
                new Card(Face.Jack, Suit.Heart),
            });
            Assert.AreEqual(3, actual.Single.Count);
            Assert.AreEqual(0, actual.Pair.Count);
            Assert.AreEqual(0, actual.ThreeOfAKind.Count);
            Assert.AreEqual(1, actual.FourOfAKind.Count);
            Assert.AreEqual(4, actual.FourOfAKind[0].Count);
            Assert.AreEqual("Ah", actual.FourOfAKind[0][0].ToString());
            Assert.AreEqual("Ad", actual.FourOfAKind[0][1].ToString());
            Assert.AreEqual("Ac", actual.FourOfAKind[0][2].ToString());
            Assert.AreEqual("As", actual.FourOfAKind[0][3].ToString());
        }

        [TestMethod]
        public void Ctor_ThreeOfAKind_ShouldReturnIt()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.Ace, Suit.Club),
                new Card(Face.King, Suit.Heart),
                new Card(Face.Queen, Suit.Heart),
                new Card(Face.Jack, Suit.Heart),
                new Card(Face.Ten, Suit.Heart),
            });
            Assert.AreEqual(4, actual.Single.Count);
            Assert.AreEqual(0, actual.Pair.Count);
            Assert.AreEqual(1, actual.ThreeOfAKind.Count);
            Assert.AreEqual("Ah", actual.ThreeOfAKind[0][0].ToString());
            Assert.AreEqual("Ad", actual.ThreeOfAKind[0][1].ToString());
            Assert.AreEqual("Ac", actual.ThreeOfAKind[0][2].ToString());
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }

        [TestMethod]
        public void Ctor_MultipleThreeOfAKind_ShouldReturnInCorrectOrder()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.Ace, Suit.Club),
                new Card(Face.King, Suit.Heart),
                new Card(Face.King, Suit.Diamond),
                new Card(Face.King, Suit.Club),
                new Card(Face.Ten, Suit.Heart),
            });
            Assert.AreEqual(1, actual.Single.Count);
            Assert.AreEqual(0, actual.Pair.Count);
            Assert.AreEqual(2, actual.ThreeOfAKind.Count);
            Assert.AreEqual(3, actual.ThreeOfAKind[0].Count);
            Assert.AreEqual("Ah", actual.ThreeOfAKind[0][0].ToString());
            Assert.AreEqual("Ad", actual.ThreeOfAKind[0][1].ToString());
            Assert.AreEqual("Ac", actual.ThreeOfAKind[0][2].ToString());
            Assert.AreEqual(3, actual.ThreeOfAKind[1].Count);
            Assert.AreEqual("Kh", actual.ThreeOfAKind[1][0].ToString());
            Assert.AreEqual("Kd", actual.ThreeOfAKind[1][1].ToString());
            Assert.AreEqual("Kc", actual.ThreeOfAKind[1][2].ToString());
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }

        [TestMethod]
        public void Ctor_Pair_ShouldReturnIt()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.Queen, Suit.Club),
                new Card(Face.Jack, Suit.Spade),
                new Card(Face.Ten, Suit.Heart),
                new Card(Face.Nine, Suit.Diamond),
                new Card(Face.Eight, Suit.Club),
            });
            Assert.AreEqual(5, actual.Single.Count);
            Assert.AreEqual(1, actual.Single[0].Count);
            Assert.AreEqual(1, actual.Single[1].Count);
            Assert.AreEqual(1, actual.Single[2].Count);
            Assert.AreEqual(1, actual.Single[3].Count);
            Assert.AreEqual(1, actual.Single[4].Count);
            Assert.AreEqual("Qc", actual.Single[0][0].ToString());
            Assert.AreEqual("Js", actual.Single[1][0].ToString());
            Assert.AreEqual(1, actual.Pair.Count);
            Assert.AreEqual(2, actual.Pair[0].Count);
            Assert.AreEqual("Ah", actual.Pair[0][0].ToString());
            Assert.AreEqual("Ad", actual.Pair[0][1].ToString());
            Assert.AreEqual(0, actual.ThreeOfAKind.Count);
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }

        [TestMethod]
        public void Ctor_MultiplePair_ShouldReturnInCorrectOrder()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.King, Suit.Heart),
                new Card(Face.King, Suit.Diamond),
                new Card(Face.Queen, Suit.Heart),
                new Card(Face.Queen, Suit.Diamond),
                new Card(Face.Jack, Suit.Heart),
            });
            Assert.AreEqual(1, actual.Single.Count);
            Assert.AreEqual(1, actual.Single[0].Count);
            Assert.AreEqual("Jh", actual.Single[0][0].ToString());
            Assert.AreEqual(3, actual.Pair.Count);
            Assert.AreEqual(2, actual.Pair[0].Count);
            Assert.AreEqual("Ah", actual.Pair[0][0].ToString());
            Assert.AreEqual("Ad", actual.Pair[0][1].ToString());
            Assert.AreEqual(2, actual.Pair[1].Count);
            Assert.AreEqual("Kh", actual.Pair[1][0].ToString());
            Assert.AreEqual("Kd", actual.Pair[1][1].ToString());
            Assert.AreEqual(2, actual.Pair[2].Count);
            Assert.AreEqual("Qh", actual.Pair[2][0].ToString());
            Assert.AreEqual("Qd", actual.Pair[2][1].ToString());
            Assert.AreEqual(0, actual.ThreeOfAKind.Count);
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }

        [TestMethod]
        public void Ctor_Singles_ShouldReturnInCorrectOrder()
        {
            var actual = new HandGrouping(SimpleCards.Take(7).ToArray());
            Assert.AreEqual(7, actual.Single.Count);
            Assert.AreEqual(1, actual.Single[0].Count);
            Assert.AreEqual(1, actual.Single[1].Count);
            Assert.AreEqual(1, actual.Single[2].Count);
            Assert.AreEqual(1, actual.Single[3].Count);
            Assert.AreEqual(1, actual.Single[4].Count);
            Assert.AreEqual(1, actual.Single[5].Count);
            Assert.AreEqual(1, actual.Single[6].Count);
            Assert.AreEqual("Ah", actual.Single[0][0].ToString());
            Assert.AreEqual("Kd", actual.Single[1][0].ToString());
            Assert.AreEqual("Qc", actual.Single[2][0].ToString());
            Assert.AreEqual("Js", actual.Single[3][0].ToString());
            Assert.AreEqual("Th", actual.Single[4][0].ToString());
            Assert.AreEqual("9d", actual.Single[5][0].ToString());
            Assert.AreEqual("8c", actual.Single[6][0].ToString());
            Assert.AreEqual(0, actual.Pair.Count);
            Assert.AreEqual(0, actual.ThreeOfAKind.Count);
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }

        [TestMethod]
        public void Ctor_FourOfAKindWithHighCards_ShouldReturnCorrectly()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.Ace, Suit.Club),
                new Card(Face.Ace, Suit.Spade),
                new Card(Face.King, Suit.Heart),
                new Card(Face.Queen, Suit.Heart),
                new Card(Face.Jack, Suit.Heart),
            });
            Assert.AreEqual(3, actual.Single.Count);
            Assert.AreEqual(1, actual.Single[0].Count);
            Assert.AreEqual(1, actual.Single[1].Count);
            Assert.AreEqual(1, actual.Single[2].Count);
            Assert.AreEqual("Kh", actual.Single[0][0].ToString());
            Assert.AreEqual("Qh", actual.Single[1][0].ToString());
            Assert.AreEqual("Jh", actual.Single[2][0].ToString());
            Assert.AreEqual(0, actual.Pair.Count);
            Assert.AreEqual(0, actual.ThreeOfAKind.Count);
            Assert.AreEqual(1, actual.FourOfAKind.Count);
            Assert.AreEqual(4, actual.FourOfAKind[0].Count);
            Assert.AreEqual("Ah", actual.FourOfAKind[0][0].ToString());
            Assert.AreEqual("Ad", actual.FourOfAKind[0][1].ToString());
            Assert.AreEqual("Ac", actual.FourOfAKind[0][2].ToString());
            Assert.AreEqual("As", actual.FourOfAKind[0][3].ToString());
        }

        [TestMethod]
        public void Ctor_FullHouseWithHighCards_ShouldReturnCorrectly()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.Ace, Suit.Club),
                new Card(Face.King, Suit.Heart),
                new Card(Face.King, Suit.Diamond),
                new Card(Face.Queen, Suit.Heart),
                new Card(Face.Jack, Suit.Heart),
            });
            Assert.AreEqual(2, actual.Single.Count);
            Assert.AreEqual(1, actual.Single[0].Count);
            Assert.AreEqual(1, actual.Single[1].Count);
            Assert.AreEqual("Qh", actual.Single[0][0].ToString());
            Assert.AreEqual("Jh", actual.Single[1][0].ToString());
            Assert.AreEqual(1, actual.Pair.Count);
            Assert.AreEqual(2, actual.Pair[0].Count);
            Assert.AreEqual("Kh", actual.Pair[0][0].ToString());
            Assert.AreEqual("Kd", actual.Pair[0][1].ToString());
            Assert.AreEqual(1, actual.ThreeOfAKind.Count);
            Assert.AreEqual(3, actual.ThreeOfAKind[0].Count);
            Assert.AreEqual("Ah", actual.ThreeOfAKind[0][0].ToString());
            Assert.AreEqual("Ad", actual.ThreeOfAKind[0][1].ToString());
            Assert.AreEqual("Ac", actual.ThreeOfAKind[0][2].ToString());
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }

        [TestMethod]
        public void Ctor_ThreeOfAKindWithHighCards_ShouldReturnCorrectly()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.Ace, Suit.Club),
                new Card(Face.King, Suit.Heart),
                new Card(Face.Queen, Suit.Heart),
                new Card(Face.Jack, Suit.Heart),
                new Card(Face.Ten, Suit.Heart),
            });
            Assert.AreEqual(4, actual.Single.Count);
            Assert.AreEqual(1, actual.Single[0].Count);
            Assert.AreEqual(1, actual.Single[1].Count);
            Assert.AreEqual(1, actual.Single[2].Count);
            Assert.AreEqual(1, actual.Single[3].Count);
            Assert.AreEqual("Kh", actual.Single[0][0].ToString());
            Assert.AreEqual("Qh", actual.Single[1][0].ToString());
            Assert.AreEqual("Jh", actual.Single[2][0].ToString());
            Assert.AreEqual("Th", actual.Single[3][0].ToString());
            Assert.AreEqual(0, actual.Pair.Count);
            Assert.AreEqual(1, actual.ThreeOfAKind.Count);
            Assert.AreEqual(3, actual.ThreeOfAKind[0].Count);
            Assert.AreEqual("Ah", actual.ThreeOfAKind[0][0].ToString());
            Assert.AreEqual("Ad", actual.ThreeOfAKind[0][1].ToString());
            Assert.AreEqual("Ac", actual.ThreeOfAKind[0][2].ToString());
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }

        [TestMethod]
        public void Ctor_TwoPairWithHighCards_ShouldReturnCorrectly()
        {
            var actual = new HandGrouping(new[]
            {
                new Card(Face.Ace, Suit.Heart),
                new Card(Face.Ace, Suit.Diamond),
                new Card(Face.King, Suit.Heart),
                new Card(Face.King, Suit.Diamond),
                new Card(Face.Queen, Suit.Heart),
                new Card(Face.Jack, Suit.Heart),
                new Card(Face.Ten, Suit.Heart),
            });
            Assert.AreEqual(3, actual.Single.Count);
            Assert.AreEqual(1, actual.Single[0].Count);
            Assert.AreEqual(1, actual.Single[1].Count);
            Assert.AreEqual(1, actual.Single[2].Count);
            Assert.AreEqual("Qh", actual.Single[0][0].ToString());
            Assert.AreEqual("Jh", actual.Single[1][0].ToString());
            Assert.AreEqual("Th", actual.Single[2][0].ToString());
            Assert.AreEqual(2, actual.Pair.Count);
            Assert.AreEqual(2, actual.Pair[0].Count);
            Assert.AreEqual("Ah", actual.Pair[0][0].ToString());
            Assert.AreEqual("Ad", actual.Pair[0][1].ToString());
            Assert.AreEqual(2, actual.Pair[1].Count);
            Assert.AreEqual("Kh", actual.Pair[1][0].ToString());
            Assert.AreEqual("Kd", actual.Pair[1][1].ToString());
            Assert.AreEqual(0, actual.ThreeOfAKind.Count);
            Assert.AreEqual(0, actual.FourOfAKind.Count);
        }
    }
}
